applepay/utils.py

```diff
@@ -66,3 +85,295 @@ def signing_time_is_valid(signature, current_time, threshold):
+    Returns:
+        string: The EC point byte string minus the uncompressed
+            byte prefix indicator
+    """
+    if not point.startswith("\x04"):
+        logger.warning("Expected uncompressed EC point.")
+        return None
+
+    return point[1:]
+
+
+def get_first_from_iterable(filter_func, iterable):
+    """Get the first filtered item from an iterable.
+
+    Args:
+        filter_func (callable): the function to filter on
+        iterable (iterable): the iterable to filter on
+    Returns:
+        object: the first filtered item from iterable or None
+            if no items matching the filter are found
+    """
+    filtered_stuff = ifilter(filter_func, iterable)
+    return next(filtered_stuff, None)
+
+
+def get_hashfunc_by_name(name, data):
+    """
+    Get a callable hashfunc by name.
+
+    This function can be used directly or with functools.partial, for example:
+
+    >>> hashfunc = functools.partial(get_hashfunc_by_name, 'sha256')
+    >>> hashfunc('sir robin').digest()
+
+    Args:
+        name (string): The string name of the desired algorithm
+        data (buffer): The buffer to hash
+    Returns:
+        callable: The callable hash function of the provided
+            algorithm updated with the data to be hashed
+    """
+    hashfunc = hashlib.new(name)
+    hashfunc.update(data)
+    return hashfunc
+
+
+def verify_signature(token):
+    """Verify the signature within an apple pay token according to
+    the #ApplePaySpec documentation."""
+
+    # We only bother to validate EC_v1 at this time.  RSA is only used
+    # for transactions from China and is not supported at this time.
+    if token['version'] != 'EC_v1':
+        logger.warning("Unsupported version {}".format(token['version']))
+        return False
+
+    # extract and decode the signature object
+    # into a CMS object
+    signature = token['signature']
+    signature_data = base64.b64decode(signature)
+    content_info = cms.ContentInfo.load(signature_data)
+    signed_data = content_info['content']
+    certificates = signed_data['certificates']
+
+    # there should only be 2 certificates present
+    if len(certificates) != 2:
+        logger.warning("Expected 2 certificates, found {}".format(len(certificates)))
+        return False
+
+    leaf_cert, intermediate_cert = get_leaf_and_intermediate_certs(certificates)
+
+    if not leaf_cert:
+        logger.warning("Leaf Certificate OID not found")
+        return False
+
+    if not intermediate_cert:
+        logger.warning("Intermediate certificate OID not found")
+        return False
+
+    root_der = open(payment.ROOT_CA_FILE, 'r').read()
+
+    # pass the der-encoded representation of the certs
+    if not valid_chain_of_trust(root_der, intermediate_cert.dump(), leaf_cert.dump()):
+        return False
+
+    # Build the signer information from the signer that matches the leaf cert
+    signer = get_first_from_iterable(
```

Comment:
The word `signer` sounds weird to me as I never heard of it. Is this short for signature?